firedrake/extrusion_numbering.pyx

```diff
@@ -0,0 +1,762 @@
+                    numpy.ndarray[PetscInt, ndim=1, mode="c"] offset):
+    """Build to DoF mapping from facets.
+
+    :arg mesh: The mesh.
+    :arg cell_nodes: numpy array mapping from cells to function space nodes.
+    :arg label: which set of facets to ask for (interior_facets or exterior_facets).
+    :arg offset: optional offset (extruded only).
+    :returns: numpy array mapping from facets to nodes in the closure
+        of the support of that facet.
+    """
+    cdef:
+        PETSc.DM dm
+        PETSc.Section cell_numbering
+        DMLabel clabel = NULL
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] facet_nodes
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] layer_extents
+        PetscInt f, p, i, j, pStart, pEnd, fStart, fEnd, point
+        PetscInt supportSize, facet, cell, ndof, dof
+        const PetscInt *renumbering
+        const PetscInt *support
+        PetscBool flg
+        bint variable, add_offset
+
+    if label not in {"interior_facets", "exterior_facets"}:
+        raise ValueError("Unsupported facet label '%s'", label)
+
+    dm = mesh._plex
+    try:
+        variable = not mesh.cell_set.constant_layers
+    except AttributeError:
+        variable = False
+
+    if variable and offset is None:
+        raise ValueError("Offset cannot be None with variable layer extents")
+
+    fStart, fEnd = dm.getHeightStratum(1)
+
+    ndof = cell_nodes.shape[1]
+
+    nfacet = dm.getStratumSize(label, 1)
+    shape = {"interior_facets": (nfacet, ndof*2),
+             "exterior_facets": (nfacet, ndof)}[label]
+
+    facet_nodes = numpy.full(shape, -1, dtype=IntType)
+
+    CHKERR(DMGetLabel(dm.dm, <char *>label, &clabel))
+    CHKERR(DMLabelCreateIndex(clabel, fStart, fEnd))
+
+    pStart, pEnd = dm.getChart()
+    CHKERR(ISGetIndices((<PETSc.IS?>mesh._plex_renumbering).iset, &renumbering))
+    cell_numbering = mesh._cell_numbering
+
+    facet = 0
+
+    if variable:
+        layer_extents = mesh.layer_extents
+    for p in range(pStart, pEnd):
+        point = renumbering[p]
+        if fStart <= point < fEnd:
+            CHKERR(DMLabelHasPoint(clabel, point, &flg))
+            if not flg:
+                # Not a facet we want.
+                continue
+
+            DMPlexGetSupportSize(dm.dm, point, &supportSize)
+            DMPlexGetSupport(dm.dm, point, &support)
+            for i in range(supportSize):
+                PetscSectionGetOffset(cell_numbering.sec, support[i], &cell)
+                for j in range(ndof):
+                    dof = cell_nodes[cell, j]
+                    if variable:
+                        # This facet iterates from higher than the
+                        # cell numbering of the cell, so we need
+                        # to add on the difference.
+                        dof += offset[j]*(layer_extents[point, 2] - layer_extents[support[i], 0])
+                    facet_nodes[facet, ndof*i + j] = dof
+            facet += 1
```

Comment:
This code, with a little bit of refactoring, would also do to replace the `dmplex.get_facet_nodes`.  And we can do away with carrying around the facet renumbering.